File: api/src/main/java/com/api/app/service/point/PointServiceImpl.java
package com.api.app.service.point;

import com.api.app.common.exception.ApiError;
import com.api.app.common.exception.ApiException;
import com.api.app.dto.request.point.PointHistoryRequest;
import com.api.app.dto.request.point.PointTransactionRequest;
import com.api.app.dto.response.point.PointBalanceResponse;
import com.api.app.dto.response.point.PointHistoryListResponse;
import com.api.app.dto.response.point.PointHistoryResponse;
import com.api.app.emum.MEM002;
import com.api.app.emum.MEM003;
import com.api.app.entity.MemberBase;
import com.api.app.entity.PointHistory;
import com.api.app.repository.member.MemberBaseMapper;
import com.api.app.repository.point.PointHistoryMapper;
import com.api.app.repository.point.PointHistoryTrxMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 포인트 서비스 구현
 *
 * @author system
 * @version 1.0
 * @since 2025-10-29
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class PointServiceImpl implements PointService {

    private final MemberBaseMapper memberBaseMapper;
    private final PointHistoryMapper pointHistoryMapper;
    private final PointHistoryTrxMapper pointHistoryTrxMapper;

    @Override
    @Transactional
    public void processPointTransaction(String email, PointTransactionRequest request) {
        log.debug("포인트 처리 시작: email={}, code={}, amount={}",
                  email, request.getPointTransactionCode(), request.getAmount());

        // 회원 조회
        MemberBase member = memberBaseMapper.selectMemberBaseByEmail(email);
        if (member == null) {
            throw new ApiException(ApiError.DATA_NOT_FOUND, "회원 정보를 찾을 수 없습니다");
        }

        // 적립 또는 사용 처리
        if (MEM002.EARN.getCode().equals(request.getPointTransactionCode())) {
            processEarn(member.getMemberNo(), request);
        } else if (MEM002.USE.getCode().equals(request.getPointTransactionCode())) {
            processUse(member.getMemberNo(), request);
        } else {
            throw new ApiException(ApiError.INVALID_PARAMETER, "잘못된 포인트 거래 코드입니다");
        }

        log.info("포인트 처리 완료: memberNo={}, code={}, amount={}",
                 member.getMemberNo(), request.getPointTransactionCode(), request.getAmount());
    }

    /**
     * 포인트 적립 처리
     */
    private void processEarn(String memberNo, PointTransactionRequest request) {
        // 유효기간 계산 (MEM003의 referenceValue1 사용)
        int validityDays = Integer.parseInt(MEM003.ETC.getReferenceValue1());
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime startDateTime = now.toLocalDate().atStartOfDay();
        LocalDateTime endDateTime = startDateTime.plusDays(validityDays);

        // 포인트 기록번호 생성
        String pointHistoryNo = pointHistoryTrxMapper.generatePointHistoryNo();

        // 포인트 내역 생성
        PointHistory pointHistory = new PointHistory();
        pointHistory.setPointHistoryNo(pointHistoryNo);
        pointHistory.setMemberNo(memberNo);
        pointHistory.setAmount(request.getAmount());
        pointHistory.setPointTransactionCode(request.getPointTransactionCode());
        pointHistory.setPointTransactionReasonCode(request.getPointTransactionReasonCode());
        pointHistory.setPointTransactionReasonNo(request.getPointTransactionReasonNo());
        pointHistory.setStartDateTime(startDateTime);
        pointHistory.setEndDateTime(endDateTime);
        pointHistory.setRemainPoint(request.getAmount());
        pointHistory.setRegistId(memberNo);
        pointHistory.setRegistDateTime(now);
        pointHistory.setModifyId(memberNo);
        pointHistory.setModifyDateTime(now);

        // 포인트 내역 등록
        int result = pointHistoryTrxMapper.insertPointHistory(pointHistory);
        if (result != 1) {
            throw new ApiException(ApiError.INTERNAL_SERVER_ERROR, "포인트 적립에 실패했습니다");
        }
    }

    /**
     * 포인트 사용 처리
     */
    private void processUse(String memberNo, PointTransactionRequest request) {
        Long useAmount = request.getAmount();
        LocalDateTime now = LocalDateTime.now();

        // 사용 가능한 포인트 조회
        PointBalanceResponse balance = pointHistoryMapper.selectPointBalance(memberNo);
        if (balance.getTotalPoint() < useAmount) {
            throw new ApiException(ApiError.INVALID_PARAMETER,
                    "사용 가능한 포인트가 부족합니다. 보유: " + balance.getTotalPoint() + ", 요청: " + useAmount);
        }

        // 종료일시가 가까운 순서로 포인트 사용
        List<PointHistory> availablePoints = pointHistoryMapper.selectAvailablePointHistory(memberNo);
        Long remainingAmount = useAmount;

        for (PointHistory earnHistory : availablePoints) {
            if (remainingAmount <= 0) {
                break;
            }

            Long currentRemain = earnHistory.getRemainPoint();
            Long deductAmount = Math.min(currentRemain, remainingAmount);

            // 적립 내역의 잔여 포인트 차감
            PointHistory updateEarn = new PointHistory();
            updateEarn.setPointHistoryNo(earnHistory.getPointHistoryNo());
            updateEarn.setRemainPoint(currentRemain - deductAmount);
            updateEarn.setModifyId(memberNo);
            updateEarn.setModifyDateTime(now);
            pointHistoryTrxMapper.updateRemainPoint(updateEarn);

            // 사용 내역 등록
            String useHistoryNo = pointHistoryTrxMapper.generatePointHistoryNo();
            PointHistory useHistory = new PointHistory();
            useHistory.setPointHistoryNo(useHistoryNo);
            useHistory.setMemberNo(memberNo);
            useHistory.setAmount(deductAmount);
            useHistory.setPointTransactionCode(request.getPointTransactionCode());
            useHistory.setPointTransactionReasonCode(request.getPointTransactionReasonCode());
            useHistory.setPointTransactionReasonNo(request.getPointTransactionReasonNo());
            useHistory.setUpperPointHistoryNo(earnHistory.getPointHistoryNo());
            useHistory.setRemainPoint(0L);
            useHistory.setRegistId(memberNo);
            useHistory.setRegistDateTime(now);
            useHistory.setModifyId(memberNo);
            useHistory.setModifyDateTime(now);
            pointHistoryTrxMapper.insertPointHistory(useHistory);

            remainingAmount -= deductAmount;
        }
    }

    @Override
    public PointBalanceResponse getPointBalance(String email) {
        log.debug("보유 포인트 조회: email={}", email);

        // 회원 조회
        MemberBase member = memberBaseMapper.selectMemberBaseByEmail(email);
        if (member == null) {
            throw new ApiException(ApiError.DATA_NOT_FOUND, "회원 정보를 찾을 수 없습니다");
        }

        // 보유 포인트 조회
        PointBalanceResponse response = pointHistoryMapper.selectPointBalance(member.getMemberNo());
        if (response == null) {
            response = new PointBalanceResponse();
            response.setTotalPoint(0L);
        }

        return response;
    }

    @Override
    public PointHistoryListResponse getPointHistoryList(String email, PointHistoryRequest request) {
        log.debug("포인트 내역 조회: email={}, page={}, size={}",
                  email, request.getPage(), request.getSize());

        // 회원 조회
        MemberBase member = memberBaseMapper.selectMemberBaseByEmail(email);
        if (member == null) {
            throw new ApiException(ApiError.DATA_NOT_FOUND, "회원 정보를 찾을 수 없습니다");
        }

        // 회원번호 설정
        request.setMemberNo(member.getMemberNo());

        // 포인트 내역 목록 조회
        List<PointHistoryResponse> content = pointHistoryMapper.selectPointHistoryList(request);
        Long totalElements = pointHistoryMapper.countPointHistory(request);

        // 응답 생성
        PointHistoryListResponse response = new PointHistoryListResponse();
        response.setContent(content);
        response.setPage(request.getPage());
        response.setSize(request.getSize());
        response.setTotalElements(totalElements);
        response.setTotalPages((int) Math.ceil((double) totalElements / request.getSize()));

        return response;
    }
}

File: api/src/main/kotlin/com/api/app/service/point/PointServiceImpl.kt
package com.api.app.service.point

import com.api.app.common.exception.ApiError
import com.api.app.common.exception.ApiException
import com.api.app.dto.request.point.PointHistoryRequest
import com.api.app.dto.request.point.PointTransactionRequest
import com.api.app.dto.response.point.PointBalanceResponse
import com.api.app.dto.response.point.PointHistoryListResponse
import com.api.app.dto.response.point.PointHistoryResponse
import com.api.app.emum.MEM002
import com.api.app.emum.MEM003
import com.api.app.entity.PointHistory
import com.api.app.repository.rodb.point.PointHistoryProjection
import com.api.app.repository.rodb.point.PointHistoryRepository
import com.api.app.repository.rwdb.point.PointHistoryTrxRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import kotlin.math.ceil

@Service
@Transactional(readOnly = true)
class PointServiceImpl(
    private val pointHistoryRepository: PointHistoryRepository,
    private val pointHistoryTrxRepository: PointHistoryTrxRepository
) : PointService {

    private val log = LoggerFactory.getLogger(this::class.java)

    @Transactional
    override fun processPointTransaction(memberNo: String, request: PointTransactionRequest) {
        log.debug("포인트 처리 시작: memberNo={}, code={}, amount={}", memberNo, request.pointTransactionCode, request.amount)

        when (request.pointTransactionCode) {
            MEM002.EARN.code -> processEarn(memberNo, request)
            MEM002.USE.code -> processUse(memberNo, request)
            else -> throw ApiException(ApiError.INVALID_PARAMETER, "잘못된 포인트 거래 코드입니다")
        }

        log.info("포인트 처리 완료: memberNo={}, code={}, amount={}", memberNo, request.pointTransactionCode, request.amount)
    }

    private fun processEarn(memberNo: String, request: PointTransactionRequest) {
        val validityDays = MEM003.ETC.referenceValue1.toInt()
        val now = LocalDateTime.now()
        val startDateTime = now.toLocalDate().atStartOfDay()
        val endDateTime = startDateTime.plusDays(validityDays.toLong())

        val pointHistory = PointHistory().apply {
            this.memberNo = memberNo
            this.amount = request.amount
            this.pointTransactionCode = request.pointTransactionCode
            this.pointTransactionReasonCode = request.pointTransactionReasonCode
            this.pointTransactionReasonNo = request.pointTransactionReasonNo
            this.startDateTime = startDateTime
            this.endDateTime = endDateTime
            this.remainPoint = request.amount
        }

        pointHistoryTrxRepository.save(pointHistory)
    }

    private fun processUse(memberNo: String, request: PointTransactionRequest) {
        val balance = pointHistoryRepository.selectPointBalance(memberNo)
        if (balance < request.amount) {
            throw ApiException(ApiError.INVALID_PARAMETER,
                "사용 가능한 포인트가 부족합니다. 보유: $balance, 요청: ${request.amount}")
        }

        val availablePoints = pointHistoryRepository.selectAvailablePointHistory(memberNo)
        var remainingAmount = request.amount

        for (earnHistory in availablePoints) {
            if (remainingAmount <= 0) break

            val currentRemain = earnHistory.remainPoint ?: 0L
            val deductAmount = minOf(currentRemain, remainingAmount)

            pointHistoryTrxRepository.updateRemainPoint(
                earnHistory.pointHistoryNo,
                currentRemain - deductAmount,
                memberNo
            )

            val useHistory = PointHistory().apply {
                this.memberNo = memberNo
                this.amount = deductAmount
                this.pointTransactionCode = request.pointTransactionCode
                this.pointTransactionReasonCode = request.pointTransactionReasonCode
                this.pointTransactionReasonNo = request.pointTransactionReasonNo
                this.upperPointHistoryNo = earnHistory.pointHistoryNo
                this.remainPoint = 0L
            }
            pointHistoryTrxRepository.save(useHistory)

            remainingAmount -= deductAmount
        }
    }

    override fun getPointBalance(memberNo: String): PointBalanceResponse {
        val balance = pointHistoryRepository.selectPointBalance(memberNo)
        return PointBalanceResponse(totalPoint = balance)
    }

    override fun getPointHistoryList(memberNo: String, request: PointHistoryRequest): PointHistoryListResponse {
        val offset = request.page.toLong() * request.size
        val content = pointHistoryRepository.selectPointHistoryList(memberNo, request.size, offset)
            .map { it.toResponse() }
        val totalElements = pointHistoryRepository.countPointHistory(memberNo)
        val totalPages = ceil(totalElements.toDouble() / request.size).toInt()

        return PointHistoryListResponse(
            content = content,
            page = request.page,
            size = request.size,
            totalElements = totalElements,
            totalPages = totalPages
        )
    }

    private fun PointHistoryProjection.toResponse() = PointHistoryResponse(
        pointHistoryNo = getPointHistoryNo(),
        amount = getAmount(),
        pointTransactionCode = getPointTransactionCode(),
        pointTransactionReasonCode = getPointTransactionReasonCode(),
        pointTransactionReasonNo = getPointTransactionReasonNo(),
        startDateTime = getStartDateTime(),
        endDateTime = getEndDateTime(),
        remainPoint = getRemainPoint(),
        createdDate = getCreatedDate()
    )
}
